sensor_msgs/include/sensor_msgs/impl/point_cloud2_iterator.hpp

```diff
@@ -272,8 +273,9 @@ PointCloud2IteratorBase<T, TT, U, C, V>::PointCloud2IteratorBase(
   int offset = set_field(cloud_msg, field_name);
 
   data_char_ = &(cloud_msg.data.front()) + offset;
-  data_ = reinterpret_cast<TT *>(data_char_);
-  data_end_ = reinterpret_cast<TT *>(&(cloud_msg.data.back()) + 1 + offset);
+  std::memcpy(&data_, &data_char_, sizeof data_char_);
```

Comment:
Oh, wow.  This took me a bit to figure out what you are doing here.  These are copies of the *pointer* from `data_char_` to `data_` (i.e. the types being copied are `**`).  That's a bit subtle; can you add a comment describing that and why we are doing it?

Also, throughout the ROS 2 codebase it looks like we prefer to use the function `sizeof(data_char_)` rather than the builtin `sizeof data_char_`; can you switch that as well?

Reply:
You're right, this copies the pointer (not the data pointed by the pointer). `std::memcpy` ensures that memory is aligned, I'll add a comment.

According to https://en.cppreference.com/w/cpp/language/sizeof `sizeof(...)` is only used for types. In this case, `data_char_` is an expression, so the second form of `sizeof` without parenthesis is preferred.

Reply:
I've added the comments in f17254bbd937fe41ba8796120d0b1127d2862a15 but kept the `sizeof` without parenthesis.

Reply:
I didn't know that; thanks for the heads up.  Looks good now, thanks!